Fix big endian serialization

chqrlie opened the pull request:
Big endian serialization was broken because:
- it partially relied on `WORDS_ENDIAN` (unconditionally undef'd in cutils.h)
- endianness was not handled at all in the bc reader.
- `bc_tag_str` was missing the `"RegExp"` string
- `lre_byte_swap()` was broken for `REOP_range` and `REOP_range32`

Modifications:
- remove `WORDS_ENDIAN`
- use `bc_put_u32()` / `bc_put_u64()` in `JS_WriteBigInt()`
- use `bc_get_u32()` / `bc_get_u64()` in `JS_ReadBigInt()`
- handle host endianness in `bc_get_u16()`, `bc_get_u32()`, `bc_get_u64()` and `JS_ReadFunctionBytecode()`
- handle optional littleEndian argument as specified in `js_dataview_getValue()` and `js_dataview_setValue()`
- fix `bc_tag_str` and `lre_byte_swap()`

chqrlie committed:
Fix big endian serialization

Big endian serialization was broken because:
- it partially relied on `WORDS_ENDIAN` (unconditionally undef'd in cutils.h)
- endianness was not handled at all in the bc reader.

Modifications:
- remove `WORDS_ENDIAN`
- use `bc_put_u32()` / `bc_put_u64()` in `JS_WriteBigInt()`
- use `bc_get_u32()` / `bc_get_u64()` in `JS_ReadBigInt()`
- handle host endianness in `bc_get_u16()`, `bc_get_u32()`, `bc_get_u64()` and
  `JS_ReadFunctionBytecode()`

- handle optional littleEndian argument as specified in
  `js_dataview_getValue()` and `js_dataview_setValue()`

chqrlie committed:
Fix big endian serialization

Big endian serialization was broken because:
- it partially relied on `WORDS_ENDIAN` (unconditionally undef'd in cutils.h)
- endianness was not handled at all in the bc reader.
- `bc_tag_str` was missing the `"RegExp"` string
- `lre_byte_swap()` was broken for `REOP_range` and `REOP_range32`

Modifications:
- remove `WORDS_ENDIAN`
- use `bc_put_u32()` / `bc_put_u64()` in `JS_WriteBigInt()`
- use `bc_get_u32()` / `bc_get_u64()` in `JS_ReadBigInt()`
- handle host endianness in `bc_get_u16()`, `bc_get_u32()`, `bc_get_u64()` and
  `JS_ReadFunctionBytecode()`
- handle optional littleEndian argument as specified in
  `js_dataview_getValue()` and `js_dataview_setValue()`
- fix `bc_tag_str` and `lre_byte_swap()`

saghul commented:
Thanks @chqrlie ! Is any test currently testing this behavior? Any chance one could be written?

chqrlie commented:
> Thanks @chqrlie ! Is any test currently testing this behavior? Any chance one could be written?

There are no big endian architectures in the current CI tests.  This actually explains why the partial implementation did not trigger any errors as all the big endian support essentially expands to nothing.

Test cases can be added to `test_bjson.js` for all cases where endianness matters (byte code, wide strings, complex regexp, typed arrays...) and running `make test` on a big endian target will be an effective test for the serialization code.

Maybe we can add `mips-unknown-linux-gnu`, or `build-linux-s390x` as in https://github.com/google/flatbuffers/pull/7707/files

saghul commented:
Yep, that would be very useful.

Do you want to give it a try?

Once you add anything to the yaml file and push the new CI will run.

chqrlie commented:
Do you mind doing this yourself? I am not familiar enough with the yaml stuff yet, I don't want to break anything.

saghul commented:
No worries! I'll take a look in the upcoming days, I'm traveling this week.

bnoordhuis reviewed (approved):
LGTM

@saghul can I go and merge this or should I wait for your CI work? Either is fine by me.

chqrlie commented:
> @saghul can I go and merge this or should I wait for your CI work? Either is fine by me.

Actually, I would like to refine this patch and clarify the methodology.  I shall post another commit later today (after I get some sleep :)

saghul commented:
> LGTM
> 
> 
> 
> @saghul can I go and merge this or should I wait for your CI work? Either is fine by me.

Land it at will, once @chqrlie is ready, I'll add the CI afterwards.

bnoordhuis commented on quickjs.c:
```diff
@@ -50421,12 +50425,9 @@ static JSValue js_dataview_getValue(JSContext *ctx,
-#ifndef WORDS_BIGENDIAN
-    is_swap ^= 1;
-#endif
+        is_swap = !JS_ToBool(ctx, argv[1]);
```
Can you walk me through this? `view.getInt32(42)` should read a big-endian value, meaning:

1. Byte-swap when system is LE
2. No byte-swap when system is BE

And vice versa, `view.getInt32(42, /*little endian*/true)` means:

1. No byte-swap when system is LE
2. Byte-swap when system is BE

It seems to me the logic should be something like this?
```c
is_le = argc > 1 && JS_ToBool(ctx, argv[1]);
is_swap = is_le ^ !is_be();
```

chqrlie replied:
@bnoordhuis: good catch! you are correct, I shall post a new commit later today.

chqrlie committed:
Fix big endian serialization

Big endian serialization was broken because:
- it partially relied on `WORDS_ENDIAN` (unconditionally undef'd in cutils.h)
- endianness was not handled at all in the bc reader.
- `bc_tag_str` was missing the `"RegExp"` string
- `lre_byte_swap()` was broken for `REOP_range` and `REOP_range32`

Modifications:
- remove `WORDS_ENDIAN`
- use `bc_put_u32()` / `bc_put_u64()` in `JS_WriteBigInt()`
- use `bc_get_u32()` / `bc_get_u64()` in `JS_ReadBigInt()`
- handle host endianness in `bc_get_u16()`, `bc_get_u32()`, `bc_get_u64()` and
  `JS_ReadFunctionBytecode()`
- handle optional littleEndian argument as specified in
  `js_dataview_getValue()` and `js_dataview_setValue()`
- fix `bc_tag_str` and `lre_byte_swap()`

bnoordhuis commented on quickjs.c:
```diff
@@ -50512,7 +50513,8 @@ static JSValue js_dataview_setValue(JSContext *ctx,
     JSTypedArray *ta;
     JSArrayBuffer *abuf;
-    int is_swap, size;
+    BOOL littleEndian, is_swap;
```
I know it's nitpicking but for consistency I would suggest calling it `little_endian`.

chqrlie replied:
I used `littleEndian` for consistency with the ECMA document, otherwise I would definitely use `little_endian`.

bnoordhuis reviewed (approved):
LGTM. Charlie, do you want me to give you commit access so you can merge it yourself?

chqrlie commented:
> LGTM. Charlie, do you want me to give you commit access so you can merge it yourself?

Thank you for proposing. I appreciate.  I guess I would also get reviewing rights?
What is your policy for reviewing before committing?

saghul commented:
Anyone can review, committee or not.

In general we open PRs for every change and it should be reviewed by at least 1 person before merging.

saghul commented:
Just sent you an invite @chqrlie !

chqrlie commented:
> No worries! I'll take a look in the upcoming days, I'm traveling this week.

I merged the new code, but we still do not have a big-endian CI target...

saghul commented:
Life happened 😅 I think I'll have time this week 💪
